### DISCUSSION TITLE
Pagination Revalidation

### DISCUSSION THREAD

**Discussion by willemliufdmg** [vercel/next.js#37584] [Help]
I was wondering if and how one would revalidate paginated pages.
E.g. take the following route: `/nieuws/[page]/index.tsx`
Revalidate is set to 5 minutes.

When I navigate to `/nieuws/1` I can see the articles. And I can see that this page gets revalidated when the set timer has been reached. However when I'm on this revalidated page and I navigate to the next page `/nieuws/2` I see an older page 2.
For example if on page 1 we have added 1 extra article. Then I would expect the last article on that page to be pushed to the top of page 2. But this doesn't happen. This is more painful when using `next/link` for SPA navigation. As that seems to never trigger the cache update. I have to manually do a hard reload of the page in order to see the updated page.

The app is deployed to Vercel.

Has anyone solved this problem? If so, could you share how you did it?

**Comment by ashishaugust22** [comment]
Hi @willemliufdmg 
So, cache using revalidate will never be updated without getting the request. Please check below equation

`revalidate: 60` // revalidate after 60 seconds is passed

Now, let's say the next request comes after 1 hour, so the revalidate will happen when that request hits your server.
So, there will be no revalidation until the request hits the server.

**Reply by willemliufdmg** [reply]
Thanks @ashishaugust22 I understand how revalidate works. I was just wondering if anyone has tried implementing pagination and how they actually fixed this problem.

**Comment by predaytor** [comment]
You can use manual headers:
```ts
res.setHeader('Cache-Control', 'public, s-maxage=10, stale-while-revalidate=59');

**Reply by willemliufdmg** [reply]
But this does not change the way how cache invalidation works right? Someone still has to visit page 2 before the actual cache is replaced by the new render. So if a lot of people visit page 1 (which is often the case). Page 1 cache is probably more fresh than page 2.
I was wondering if there is a solution for SSG pages with pagination to update collectively without having to revalidate each "possible" page one revalidate call at a time.

**Reply by predaytor** [reply]
"If you use Cache-Control to cache in the user browsers you will not have a way to invalidate that cache after an action is run, so the user will still receive cached data. If you want to be able to purge the cache you need to use Cache-Control to only cache in a CDN and use a CDN which allow you to purge the cache manually by sending some request to an API, like Cloudflare or Fastly.". 
Quote from @sergiodxa article on https://sergiodxa.com

I don't know how to do this for all pages. How do you handle this with the SSG approach? I'm using SSR and using the following query parameters like: `/?page=1`.

**Reply by willemliufdmg** [reply]
With SSG the headers are automatically set for you by Next.js. So a page is rendered and cached. Then when revalidate timer has been reached and a user visits it after that timer then a new render will take place. That user however will see the version of the page still in the cache. Not yet the new page. This behaviour is standard.

However my initial question was if anyone has solved the problem regarding pagination.
Because with SSG for example my first user which navigates to page 1, 2, 3 will cause pages 1, 2 and 3 to be rendered and put to the cache.
Lets say now a new "item" has been published and it should show at the top of page 1.
User 2 goes to page 1 and sees the old cached page but triggers the re-render of page 1.
User 3 goes to page 1 and sees the new page with the new "item" at the top. User 3 navigates to page 2 and gets served the old cached version of page 2. This means that the last item at the bottom of page 1 which should've been on top of page 2 is now missing.
